snapfaas/src/vm.rs

```diff
@@ -532,7 +567,133 @@ impl Vm {
+                        let data = wb.data.as_ref();
+                        if newblob.write_all(data).is_ok() {
+                            syscalls::BlobResponse {
+                                success: true,
+                                fd: wb.fd,
+                                data: Vec::new(),
+                            }
+                        } else {
+                            syscalls::BlobResponse {
+                                success: false,
+                                fd: wb.fd,
+                                data: Vec::from("Failed to write"),
+                            }
+                        }
+                    } else {
+                        syscalls::BlobResponse {
+                            success: false,
+                            fd: wb.fd,
+                            data: Vec::from("Blob doesn't exist"),
+                        }
+                    };
+                    self.send_into_vm(result.encode_to_vec())?;
+                },
+                Some(SC::FinalizeBlob(fb)) => {
+                    let result = if let Some(mut newblob) = self.create_blobs.remove(&fb.fd) {
+                        let blob = newblob.write_all(&fb.data).and_then(|_| self.blobstore.save(newblob))?;
+                        syscalls::BlobResponse {
+                            success: true,
+                            fd: fb.fd,
+                            data: Vec::from(blob.name),
+                        }
+                    } else {
+                        syscalls::BlobResponse {
+                            success: false,
+                            fd: fb.fd,
+                            data: Vec::from("Blob doesn't exist"),
+                        }
+                    };
+                    self.send_into_vm(result.encode_to_vec())?;
+                },
+                Some(SC::OpenBlob(ob)) => {
+                    let result = if let Ok(file) = self.blobstore.open(ob.name) {
+                        self.max_blob_id += 1;
+                        self.blobs.insert(self.max_blob_id, file);
+                        syscalls::BlobResponse {
+                            success: true,
+                            fd: self.max_blob_id,
+                            data: Vec::new(),
+                        }
+                    } else {
+                        syscalls::BlobResponse {
+                            success: false,
+                            fd: 0,
+                            data: Vec::new(),
+                        }
+                    };
+                    self.send_into_vm(result.encode_to_vec())?;
+                },
+                Some(SC::ReadBlob(rb)) => {
+                    let result = if let Some(file) = self.blobs.get_mut(&rb.fd) {
+                        let mut buf = Vec::from([0; 4096]);
+                        let limit = std::cmp::min(rb.length.unwrap_or(4096), 4096) as usize;
```

Comment:
My interpretation is that the 4KB limit means that the VM-side syscall `read_blob` handles calling this syscall multiple times for data larger than 4K?

Reply:
Yes, it needs to. In fact, read is not guaranteed to fill all 4k, it might return with less, just like a unix read system call (I think the right semantics is that it is implementation specific what this limit is, and userspace just needs to handle getting back less than it asked for)